pkg/trace/api/responses.go

```diff
@@ -68,16 +69,34 @@ func httpOK(w http.ResponseWriter) {
 	io.WriteString(w, "OK\n")
 }
 
+type writeCounter struct {
+	w io.Writer
+	n uint64
+}
+
+func newWriteCounter(w io.Writer) *writeCounter {
+	return &writeCounter{w: w}
+}
+
+func (wc *writeCounter) Write(p []byte) (n int, err error) {
+	atomic.AddUint64(&wc.n, uint64(len(p)))
```

Comment:
There is no need to use the funcs from the `atomic` package to read/write `wc.n`. The variable is only ever accessed by a single goroutine.

Reply:
My reasoning was that the performance to using `atomic` is negligible, and the advantage of it is that the `writeCounter` becomes usable in concurrent situations. Although, I now realise that there are may not be many situations where concurrent writing is a good idea. Do you think that it's a precautionary method that is not needed? Or better to have?

Reply:
I believe it's best to avoid introducing the code handling concurrency where it is not necessary:
- The readers may get confused as to why this is necessary (I'd certainly spend some time figuring things out if I saw this), those unfamiliar with the Go memory model may end up using this code as an example.
- There's a performance penalty. As you pointed out, it's likely negligible in this particular case.

> Although, I now realise that there may not be many situations where concurrent writing is a good idea.

Right. Also, this code is unexported, so we are aware of all of its usages and we know none of them have concurrent writes.

Reply:
Those are good points. But for some reason, I'm still not comfortable throwing an `io.Writer` at some external code and not accounting for this possibility. I don't want to risk any races and would like to keep the atomic version. Do you feel strongly about it? I don't, but I'm not fully confident that it's safe to make assumption about other implementations.

Reply:
Please have a look at pretty much any other fairly generic implementation of an `io.Writer` or an `io.Reader`, including those in the standard library: `gzip.Writer`, `bufio.Writer`, `http.MaxBytesReader`, ... - none of them bother with concurrency.

The assumption the reader/writer is not going to be accessed concurrently is a very reasonable one, too. There are very few scenarios where concurrent writes or reads would make sense.

> I'm still not comfortable throwing an io.Writer at some external code and not accounting for this possibility

Well, you're not throwing this io.Writer at some unknown external code. The only user of this code is `json.Encoder`.

Reply:
Ok, fair enough. Resolving.

Reply:
Actually, one last thing. What about the situation where `N()` might be called / accessed while the writer is still being written to? This is definitely a possibility, especially with HTTP clients.

Reply:
If you had that scenario, then you would indeed have to implement some sort of synchronization. But since you don't have it and this isn't even an exported name...

I mean, you can use this same argument (the code can hypothetically be used concurrently) to make a point that *all* code should be safe for concurrent use, can't you?

Reply:
I'm choosing to keep the synchronisation for situations where one might access `N` during a write.